File: java_version/array_and_strings/most_frequent_char/most_frequent_char.java
package array_and_strings.most_frequent_char;
// most frequent char

// Write a function, mostFrequentChar, that takes in a string as an argument. 
// The function should return the most frequent character of the string. 
// If there are ties, return the character that appears earlier in the string.

import java.util.Collections;
import java.util.HashMap;
import java.util.Map;

// You can assume that the input string is non-empty.
public class most_frequent_char {

    public static char mostFrequentChar(String s) {
        HashMap<Character, Integer> hashMap = new HashMap<>();
        for (char letter : s.toCharArray()) {
            hashMap.merge(letter, 1, Integer::sum);
        }
        return Collections.max(hashMap.entrySet(), Map.Entry.comparingByValue()).getKey();

    }

    public static void main(String[] args) {
        System.out.println(mostFrequentChar("bookeeper")); // -> 'e'
        System.out.println(mostFrequentChar("david")); // -> 'd'
        System.out.println(mostFrequentChar("abby")); // -> 'b'
        System.out.println(mostFrequentChar("mississippi")); // -> 'i'
        System.out.println(mostFrequentChar("potato")); // -> 'o'
        System.out.println(mostFrequentChar("eleventennine")); // -> 'e'
        System.out.println(mostFrequentChar("riverbed")); // -> 'r'
    }

}

File: python_version/array_and_strings/most_frequent_char/most_frequent_char.py
# most frequent char
# Write a function, mostFrequentChar, that takes in a string as an argument.
# The function should return the most frequent character of the string.
# If there are ties, return the character that appears earlier in the string.

# You can assume that the input string is non-empty.

def most_frequent_char(s):
    hsh = {}
    for i in s:
        if i in hsh:
            hsh[i] += 1
        else:
            hsh[i] = 1
    maxVal = None
    for char in hsh:
        if maxVal == None or hsh[char] > hsh[maxVal]:
            maxVal = char
    return maxVal


print(most_frequent_char('bookeeper'))  # -> 'e'
print(most_frequent_char('david'))  # -> 'd'
print(most_frequent_char('abby'))  # -> 'b'
print(most_frequent_char('mississippi'))  # -> 'i'
print(most_frequent_char('potato'))  # -> 'o'
print(most_frequent_char('eleventennine'))  # -> 'e'
print(most_frequent_char('riverbed'))  # -> 'r'
